[fix](metrics) avoid duplicated # TYPE definitions in Prometheus exporter #57788

Zhaowei-Zhang-27 opened the pull request:
### What problem does this PR solve?
**Issue Number:** close #57788 

**Related PR:** None

**Problem Summary:**
This PR fixes a bug where Prometheus metrics exported duplicated `# TYPE` definition lines. The original logic relied on comparing the `group_name` of adjacent metric entities. However, because metric entities are not strictly ordered within the `entity_metrics_by_types` container, this "adjacent-comparison" approach failed when entities of the same type were separated, resulting in redundant `# TYPE` lines that violate OpenMetrics specifications.

**Key Changes:**
* **Global Deduplication**: Introduced `std::unordered_set<std::string> exported_types` to maintain a global state of already-exported metric names during the output phase.
* **Algorithm Efficiency**: Replaced the local check with a hash-set based lookup, ensuring each `# TYPE` line is printed exactly once with an average time complexity of $O(1)$.
* **Robustness**: Ensured that the metrics output remains valid regardless of the internal storage order of metric entities.



### Release note
None

### Check List (For Author)

* [x] Test
* [x] Regression test (Verified via GitHub Actions CI)
* [x] Unit Test (Verified with backend unit tests related to metrics)
* [ ] Manual test (add detailed scripts or steps below)
* [ ] No need to test or manual test. Explain why:

### Behavior changed:
* [x] No.
* [ ] Yes.

### Does this need documentation?
* [x] No.
* [ ] Yes.

### Check List (For Reviewer who merge this PR)
* [ ] Confirm the release note
* [ ] Commission test cases
* [ ] Confirm document
* [ ] Add branch pick label

<details>
<summary>中文修复逻辑说明</summary>

### 修复逻辑说明：
针对 Prometheus 指标输出中存在重复 `# TYPE` 和 `# HELP` 行的问题（如 `doris_be_engine_requests_total`），本次 PR 进行了以下改进：

1. **引入去重状态机**：在 `to_prometheus` 方法中局部维护一个 `std::unordered_set<std::string>`，用于记录当前请求中已导出的指标元数据。
2. **条件式头部输出**：在遍历各指标实体时，逻辑会先检索该指标名是否已存在于集合中。仅在首次发现时，才按规范输出相应的类型声明和帮助文档。
3. **协议合规化**：此改动消除了 Prometheus 抓取时的 "duplicate metric family" 警告，确保了指标暴露接口严格遵循 OpenMetrics / Prometheus 文本协议标准。

</details>

Zhaowei-Zhang-27 committed:
[fix](metrics) avoid duplicated # TYPE definitions in Prometheus exporter #57788

gavinchou commented on be/src/util/metrics.cpp:
```diff
@@ -345,13 +346,12 @@ std::string MetricRegistry::to_prometheus(bool with_tablet_metrics) const {
     // Output
     std::stringstream ss;
-    std::string last_group_name;
+    std::unordered_set<std::string> exported_types;
```
please add a UT to reproduce issue #57788 and make sure this PR fix the issue.

Zhaowei-Zhang-27 committed:
Add unit test for duplicate TYPE line fix
- Add PrometheusDuplicateTypeFix test to verify the fix
- Test creates interleaved metrics to reproduce the issue
- Verify TYPE line appears only once for each metric type

Zhaowei-Zhang-27 committed:
Fix code formatting
